File: apps/write-8-minutes/ios/Anky/Core/Level/LevelSyncClient.swift
import Foundation
#if SWIFT_PACKAGE
import AnkyProtocol
#endif

/// Status mirror of the server's level ledger (`GET /level/status`).
struct LevelServerStatus: Codable, Equatable {
    let totalSeconds: Int
    let level: Int
    let secondsIntoLevel: Int
    let secondsRequired: Int
    let percent: Double
    let nextLevel: Int
    let nextPaintingPhase: LevelPaintingPhase
    let nextPaintingTitle: String?
    let nextPalette: [String]?
    let pendingCeremonyLevel: Int?
}

/// Reports sealed sessions (hash + seconds only, never writing) to the server
/// ledger and reads level status back. Same signing plumbing as MirrorClient.
struct LevelSyncClient {
    let baseURL: URL
    var session: URLSession = .shared

    init(
        baseURL: URL? = URL(string: MirrorConfiguration.currentBaseURL()),
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL ?? URL(string: MirrorConfiguration.defaultBaseURL)!
        self.session = session
    }

    enum LevelSyncError: Error, Equatable {
        case invalidResponse
        case server(statusCode: Int)
    }

    @discardableResult
    func reportSessions(
        _ sessions: [LevelUnreportedSession],
        identity: WriterIdentity
    ) async throws -> LevelServerStatus {
        struct ReportBody: Encodable {
            let sessions: [LevelUnreportedSession]
        }
        let body = try JSONEncoder().encode(ReportBody(sessions: sessions))
        var request = try signedRequest(path: "level/sessions", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let (data, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
        struct ReportEnvelope: Decodable {
            let status: LevelServerStatus
        }
        guard let envelope = try? JSONDecoder().decode(ReportEnvelope.self, from: data) else {
            throw LevelSyncError.invalidResponse
        }
        return envelope.status
    }

    func fetchStatus(identity: WriterIdentity) async throws -> LevelServerStatus {
        var request = try signedRequest(path: "level/status", body: Data(), identity: identity)
        request.httpMethod = "GET"

        let (data, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
        struct StatusEnvelope: Decodable {
            let status: LevelServerStatus
        }
        guard let envelope = try? JSONDecoder().decode(StatusEnvelope.self, from: data) else {
            throw LevelSyncError.invalidResponse
        }
        return envelope.status
    }

    /// Asks the server to distill + paint the package for `level`. The text
    /// is the writing since the last level-up — transient on the server,
    /// distilled once, forgotten. Returns the server's phase for the level.
    @discardableResult
    func prepare(level: Int, text: String, identity: WriterIdentity) async throws -> String {
        struct PrepareBody: Encodable {
            let level: Int
            let text: String
        }
        let body = try JSONEncoder().encode(PrepareBody(level: level, text: text))
        var request = try signedRequest(path: "level/prepare", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let (data, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
        struct PrepareEnvelope: Decodable {
            let phase: String?
        }
        let envelope = try? JSONDecoder().decode(PrepareEnvelope.self, from: data)
        return envelope?.phase ?? "generationPending"
    }

    /// Tells the ledger the unveiling was witnessed. Idempotent server-side.
    func reportCeremonyShown(level: Int, identity: WriterIdentity) async throws {
        struct CeremonyBody: Encodable {
            let level: Int
        }
        let body = try JSONEncoder().encode(CeremonyBody(level: level))
        var request = try signedRequest(path: "level/ceremony-shown", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (_, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
    }

    /// Fetches one file of a level's painting package (signed empty body).
    func fetchAsset(level: Int, file: String, identity: WriterIdentity) async throws -> Data {
        var request = try signedRequest(
            path: "level/assets/\(level)/\(file)",
            body: Data(),
            identity: identity
        )
        request.httpMethod = "GET"
        let (data, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
        return data
    }

    /// Analytics only (phase-2 §2): the emergency breath completed. One
    /// signed event, nothing stored server-side, never surfaced to anyone.
    func reportEmergencyUnlock(identity: WriterIdentity) async throws {
        let body = Data("{}".utf8)
        var request = try signedRequest(path: "events/emergency-unlock", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (_, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
    }

    // MARK: Subscription truth and the funnel

    /// The server's answer after an identify: its webhook-maintained
    /// entitlement for this wallet. `expiresAtMs` nil is a valid entitled
    /// state (open-ended promotional or lifetime grants).
    struct SubscriptionServerState: Codable, Equatable {
        let entitled: Bool
        let productId: String?
        let expiresAtMs: Int64?
        let store: String?
        let periodType: String?
    }

    /// Tells the mirror that this wallet is the RevenueCat appUserID. The
    /// EIP-712 headers prove the wallet; RevenueCat webhooks carry the
    /// entitlement truth server-side. The server rejects any attempt to
    /// attach a different appUserID to the authenticated account.
    @discardableResult
    func identifySubscription(identity: WriterIdentity) async throws -> SubscriptionServerState {
        struct IdentifyBody: Encodable {
            let appUserId: String
        }
        let body = try JSONEncoder().encode(IdentifyBody(appUserId: identity.address))
        var request = try signedRequest(path: "subscription/identify", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (data, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
        guard let state = try? JSONDecoder().decode(SubscriptionServerState.self, from: data) else {
            throw LevelSyncError.invalidResponse
        }
        return state
    }

    /// One signed funnel event (phase-3 §6). Whitelisted names server-side;
    /// hashed account, event, origin — never anything about writing.
    func reportFunnelEvent(
        _ event: String,
        origin: String? = nil,
        identity: WriterIdentity
    ) async throws {
        struct FunnelBody: Encodable {
            let event: String
            let origin: String?
        }
        let body = try JSONEncoder().encode(FunnelBody(event: event, origin: origin))
        var request = try signedRequest(path: "events/funnel", body: body, identity: identity)
        request.httpMethod = "POST"
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (_, response) = try await session.data(for: request)
        try Self.ensureSuccess(response)
    }

    /// Drains the unreported queue: posts a batch, marks acknowledged hashes
    /// reported, and adopts the server total if it is ahead of this install.
    static func flushUnreported(
        store: LevelProgressStore = LevelProgressStore(),
        identityStore: WriterIdentityStore = WriterIdentityStore(),
        client: LevelSyncClient = LevelSyncClient()
    ) async {
        let pending = store.unreportedSessions()
        guard !pending.isEmpty, let identity = try? identityStore.loadOrCreate() else {
            return
        }
        guard let status = try? await client.reportSessions(pending, identity: identity) else {
            return
        }
        store.markReported(hashes: pending.map(\.hash))
        store.adoptServerTotalIfHigher(status.totalSeconds)
    }

    private func signedRequest(path: String, body: Data, identity: WriterIdentity) throws -> URLRequest {
        let signed = try AnkyPostSigner.sign(body: body, identity: identity)
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.setValue(signed.identityVersion, forHTTPHeaderField: "X-Anky-Identity-Version")
        request.setValue(signed.accountId, forHTTPHeaderField: "X-Anky-Account")
        request.setValue(signed.signatureType, forHTTPHeaderField: "X-Anky-Signature-Type")
        request.setValue(signed.signature, forHTTPHeaderField: "X-Anky-Signature")
        request.setValue(signed.requestTime, forHTTPHeaderField: "X-Anky-Request-Time")
        request.setValue(signed.client, forHTTPHeaderField: "X-Anky-Client")
        return request
    }

    private static func ensureSuccess(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else {
            throw LevelSyncError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw LevelSyncError.server(statusCode: http.statusCode)
        }
    }
}

/// Fire-and-forget funnel reporting (phase-3 §6). Analytics never block or
/// break the practice: failures are silently dropped, the app never waits.
enum AnkyFunnel {
    static let ceremonyOneShown = "ceremony_1_shown"
    static let boundaryReached = "boundary_reached"
    static let veilTapped = "veil_tapped"
    static let paywallShown = "paywall_shown"
    static let trialStarted = "trial_started"
    static let subscribed = "subscribed"
    static let restored = "restored"
    static let lapsed = "lapsed"

    static func report(
        _ event: String,
        origin: String? = nil,
        identityStore: WriterIdentityStore = WriterIdentityStore(),
        client: LevelSyncClient = LevelSyncClient()
    ) {
        Task.detached(priority: .utility) {
            guard let identity = try? identityStore.loadOrCreate() else { return }
            try? await client.reportFunnelEvent(event, origin: origin, identity: identity)
        }
    }
}

File: apps/android/app/src/main/java/inc/anky/android/core/level/LevelSyncClient.kt
package inc.anky.android.core.level

import inc.anky.android.core.identity.WriterIdentity
import java.io.IOException
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject

/** Status mirror of the server's level ledger (`GET /level/status`). */
data class LevelServerStatus(
    val totalSeconds: Long,
    val level: Int,
    val secondsIntoLevel: Long,
    val secondsRequired: Long,
    val percent: Double,
    val nextLevel: Int,
    val nextPaintingPhase: LevelPaintingPhase,
    val nextPaintingTitle: String?,
    val nextPalette: List<String>?,
    val pendingCeremonyLevel: Int?,
)

/**
 * The server's answer after an identify: its webhook-maintained entitlement
 * for this wallet. `expiresAtMs` null is a valid entitled state (open-ended
 * promotional or lifetime grants).
 */
data class SubscriptionServerState(
    val entitled: Boolean,
    val productId: String?,
    val expiresAtMs: Long?,
    val store: String?,
    val periodType: String?,
)

sealed class LevelSyncError(message: String) : IOException(message) {
    object InvalidResponse : LevelSyncError("INVALID_RESPONSE") {
        private fun readResolve(): Any = InvalidResponse
    }

    data class Server(val statusCode: Int) : LevelSyncError("SERVER_$statusCode")
}

/** The network surface the painting machinery depends on (fakeable in tests). */
interface LevelSyncing {
    suspend fun reportSessions(sessions: List<LevelUnreportedSession>, identity: WriterIdentity): LevelServerStatus
    suspend fun fetchStatus(identity: WriterIdentity): LevelServerStatus
    suspend fun prepare(level: Int, text: String, identity: WriterIdentity): String
    suspend fun reportCeremonyShown(level: Int, identity: WriterIdentity)
    suspend fun fetchAsset(level: Int, file: String, identity: WriterIdentity): ByteArray
    suspend fun reportEmergencyUnlock(identity: WriterIdentity)
    suspend fun identifySubscription(identity: WriterIdentity): SubscriptionServerState
    suspend fun reportFunnelEvent(event: String, origin: String? = null, identity: WriterIdentity)
}

/**
 * Reports sealed sessions (hash + seconds only, never writing) to the server
 * ledger and reads level status back. Same signing plumbing as MirrorClient.
 */
class LevelSyncClient(
    baseUrl: String,
    private val client: OkHttpClient = OkHttpClient(),
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
) : LevelSyncing {
    private val base = baseUrl.trimEnd('/')

    override suspend fun reportSessions(
        sessions: List<LevelUnreportedSession>,
        identity: WriterIdentity,
    ): LevelServerStatus {
        val body = JSONObject()
            .put(
                "sessions",
                JSONArray().also { array ->
                    sessions.forEach { session ->
                        array.put(
                            JSONObject()
                                .put("hash", session.hash)
                                .put("seconds", session.seconds)
                                .put("sealedAtMs", session.sealedAtMs),
                        )
                    }
                },
            )
            .toString()
            .toByteArray(Charsets.UTF_8)
        val data = post("level/sessions", body, identity)
        return data.toJsonObject()?.optJSONObject("status")?.toLevelServerStatus()
            ?: throw LevelSyncError.InvalidResponse
    }

    override suspend fun fetchStatus(identity: WriterIdentity): LevelServerStatus {
        val data = get("level/status", identity)
        return data.toJsonObject()?.optJSONObject("status")?.toLevelServerStatus()
            ?: throw LevelSyncError.InvalidResponse
    }

    /**
     * Asks the server to distill + paint the package for `level`. The text
     * is the writing since the last level-up — transient on the server,
     * distilled once, forgotten. Returns the server's phase for the level.
     */
    override suspend fun prepare(level: Int, text: String, identity: WriterIdentity): String {
        val body = JSONObject()
            .put("level", level)
            .put("text", text)
            .toString()
            .toByteArray(Charsets.UTF_8)
        val data = post("level/prepare", body, identity)
        val phase = data.toJsonObject()?.optString("phase")?.takeIf { it.isNotBlank() }
        return phase ?: "generationPending"
    }

    /** Tells the ledger the unveiling was witnessed. Idempotent server-side. */
    override suspend fun reportCeremonyShown(level: Int, identity: WriterIdentity) {
        val body = JSONObject().put("level", level).toString().toByteArray(Charsets.UTF_8)
        post("level/ceremony-shown", body, identity)
    }

    /** Fetches one file of a level's painting package (signed empty body). */
    override suspend fun fetchAsset(level: Int, file: String, identity: WriterIdentity): ByteArray =
        get("level/assets/$level/$file", identity)

    /**
     * Analytics only (phase-2 §2): the emergency breath completed. One
     * signed event, nothing stored server-side, never surfaced to anyone.
     */
    override suspend fun reportEmergencyUnlock(identity: WriterIdentity) {
        post("events/emergency-unlock", "{}".toByteArray(Charsets.UTF_8), identity)
    }

    /**
     * Tells the mirror that this wallet is the RevenueCat appUserID. The
     * EIP-712 headers prove the wallet; RevenueCat webhooks carry the
     * entitlement truth server-side. The server rejects any attempt to
     * attach a different appUserID to the authenticated account.
     */
    override suspend fun identifySubscription(identity: WriterIdentity): SubscriptionServerState {
        val body = JSONObject().put("appUserId", identity.address).toString().toByteArray(Charsets.UTF_8)
        val data = post("subscription/identify", body, identity)
        val json = data.toJsonObject() ?: throw LevelSyncError.InvalidResponse
        if (!json.has("entitled")) throw LevelSyncError.InvalidResponse
        return SubscriptionServerState(
            entitled = json.getBoolean("entitled"),
            productId = json.optStringOrNull("productId"),
            expiresAtMs = if (json.has("expiresAtMs") && !json.isNull("expiresAtMs")) json.getLong("expiresAtMs") else null,
            store = json.optStringOrNull("store"),
            periodType = json.optStringOrNull("periodType"),
        )
    }

    /**
     * One signed funnel event (phase-3 §6). Whitelisted names server-side;
     * hashed account, event, origin — never anything about writing.
     */
    override suspend fun reportFunnelEvent(event: String, origin: String?, identity: WriterIdentity) {
        val body = JSONObject()
            .put("event", event)
            .apply { origin?.let { put("origin", it) } }
            .toString()
            .toByteArray(Charsets.UTF_8)
        post("events/funnel", body, identity)
    }

    private suspend fun post(path: String, body: ByteArray, identity: WriterIdentity): ByteArray =
        execute(
            SignedLevelRequests.sign(Request.Builder(), body, identity)
                .url(url(path))
                .post(body.toRequestBody("application/json".toMediaType()))
                .header("Content-Type", "application/json"),
        )

    private suspend fun get(path: String, identity: WriterIdentity): ByteArray =
        execute(
            SignedLevelRequests.sign(Request.Builder(), ByteArray(0), identity)
                .url(url(path))
                .get(),
        )

    private fun url(path: String): okhttp3.HttpUrl =
        "$base/$path".toHttpUrlOrNull() ?: throw LevelSyncError.InvalidResponse

    private suspend fun execute(builder: Request.Builder): ByteArray =
        withContext(ioDispatcher) {
            client.newCall(builder.build()).execute().use { response ->
                ensureSuccess(response)
                response.body?.bytes() ?: ByteArray(0)
            }
        }

    private fun ensureSuccess(response: Response) {
        if (response.code !in 200..299) {
            throw LevelSyncError.Server(statusCode = response.code)
        }
    }

    companion object {
        /**
         * Drains the unreported queue: posts a batch, marks acknowledged
         * hashes reported, and adopts the server total if it is ahead of
         * this install. Failures leave the queue intact for the next pass.
         */
        suspend fun flushUnreported(
            store: LevelProgressStore,
            identity: WriterIdentity?,
            client: LevelSyncing,
        ) {
            val pending = store.unreportedSessions()
            if (pending.isEmpty() || identity == null) return
            val status = runCatching { client.reportSessions(pending, identity) }.getOrNull() ?: return
            store.markReported(pending.map { it.hash })
            store.adoptServerTotalIfHigher(status.totalSeconds)
        }
    }
}

/**
 * Fire-and-forget funnel reporting (phase-3 §6). Analytics never block or
 * break the practice: failures are silently dropped, the app never waits.
 */
class AnkyFunnel(
    private val client: LevelSyncing,
    private val identityProvider: suspend () -> WriterIdentity?,
    private val scope: CoroutineScope,
) {
    fun report(event: String, origin: String? = null) {
        scope.launch {
            val identity = runCatching { identityProvider() }.getOrNull() ?: return@launch
            runCatching { client.reportFunnelEvent(event, origin, identity) }
        }
    }

    companion object {
        const val CeremonyOneShown = "ceremony_1_shown"
        const val BoundaryReached = "boundary_reached"
        const val VeilTapped = "veil_tapped"
        const val PaywallShown = "paywall_shown"
        const val TrialStarted = "trial_started"
        const val Subscribed = "subscribed"
        const val Restored = "restored"
        const val Lapsed = "lapsed"
    }
}

private fun ByteArray.toJsonObject(): JSONObject? =
    runCatching { JSONObject(toString(Charsets.UTF_8)) }.getOrNull()

private fun JSONObject.optStringOrNull(name: String): String? =
    if (has(name) && !isNull(name)) getString(name).takeIf { it.isNotEmpty() } else null

private fun JSONObject.toLevelServerStatus(): LevelServerStatus? {
    val phase = LevelPaintingPhase.fromRawValue(optString("nextPaintingPhase")) ?: return null
    if (!has("totalSeconds") || !has("level")) return null
    val paletteArray = optJSONArray("nextPalette")
    val palette = paletteArray?.let { array ->
        (0 until array.length()).map { array.getString(it) }
    }
    return LevelServerStatus(
        totalSeconds = getLong("totalSeconds"),
        level = getInt("level"),
        secondsIntoLevel = getLong("secondsIntoLevel"),
        secondsRequired = getLong("secondsRequired"),
        percent = getDouble("percent"),
        nextLevel = getInt("nextLevel"),
        nextPaintingPhase = phase,
        nextPaintingTitle = optStringOrNull("nextPaintingTitle"),
        nextPalette = palette,
        pendingCeremonyLevel = if (has("pendingCeremonyLevel") && !isNull("pendingCeremonyLevel")) {
            getInt("pendingCeremonyLevel")
        } else {
            null
        },
    )
}
